### PR TITLE
Return multiple counts

### PR DESCRIPTION
**Pull request by AndrewTorrance**
# Motivation and Context
Requirment is to display tab counts only when the ru_ref_filter is in operation

# What has changed

Get count changed to go via get_tab_counts which returns a dictionary. This in turn calls 2 secure message endpoints. One for if the ru ref filter is in use and another if not. The reason being that the call that gets 4 counts is more expensive than the one that returns 1. Tab titles passed into html via a dictionary so that highlight selection logic is eased when the actual title varies. Also fixed up date tests that failed only on 1st January 2020 by removing duplicate tests
Unit tests added that validate correct count shown on correct tab
This change is dependant on a corresponding change in secure message 

# How to test?

Run unit tests. 
Create several conversations regarding 1 ru using frontstage and response ops. Validate that normally no counts shown. Then add an ru ref filter and validate counts are displayed

# Links
https://trello.com/c/3Bj1gYTy

# Screenshots (if appropriate):

### CODE REVIEW

File: response_operations_ui/controllers/message_controllers.py [hunk 1]
```diff
@@ -53,19 +69,37 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
         raise ApiError(response)
 
     logger.info("Count successful")
+
     try:
-        return response.json()['total']
+        totals = response.json()['totals']
+
+        # Secure Message uses different identifiers to the tab names used in the ui, this translates the names
+        totals['initial'] = totals.pop('new_respondent_conversations')
```
Comment (insacuri) [inline response_operations_ui/controllers/message_controllers.py]: I think this might need to change.  You currently get the totals from the response,  remove new_respondent_conversations and my_conversations and add 'my messages', initital and current to it then return it.
What happens if total contains extra stuff (because we made a future change)?  This can be written in a way so that we can be more specific about what we want (unless we really do want everything, but I'm not so sure about implicitly taking everything)

It could look like 
```python
totals = response.json()['totals']
#not sure about the variable name >_<
ui_totals['initial'] = totals.get('new_respondent_conversations')
ui_totals['my messages'] = totals.get('my_conversations')
ui_totals['current'] = totals[conversation_tab]
```

File: response_operations_ui/controllers/message_controllers.py [hunk 2]
```diff
@@ -53,19 +69,37 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
...
+        totals['my messages'] = totals.pop('my_conversations')
+
+        totals['current'] = totals[conversation_tab]
+
+        return totals
     except KeyError:
-        logger.exception("Response was successful but didn't contain a 'total' key")
+        logger.exception("Response was successful but didn't contain a 'totals' key")
```
Comment (insacuri) [inline response_operations_ui/controllers/message_controllers.py]: A keyError isn't just a missing totals key.  It could also be a missing key in the totals dictionary.  This needs to be made more clear

File: response_operations_ui/views/messages.py [hunk 3]
```diff
@@ -231,6 +231,7 @@ def view_selected_survey(selected_survey):
     displayed_short_name = format_short_name(selected_survey)
     session['messages_survey_selection'] = selected_survey
     breadcrumbs = [{"text": displayed_short_name + " Messages"}]
+    session['totals'] = {}
```
Comment (insacuri) [inline response_operations_ui/views/messages.py]: Why's this here?  Not sure where it's used

File: response_operations_ui/controllers/message_controllers.py [hunk 4]
```diff
@@ -40,11 +40,27 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
...
+        logger.exception("Thread count failed")
+        raise ApiError(response)
 
-    url = f'{current_app.config["SECURE_MESSAGE_URL"]}/messages/count'
+    logger.info("Count successful")
+    try:
+        return response.json()['total']
+    except KeyError:
+        logger.exception("Response was successful but didn't contain a 'total' key")
+        raise NoMessagesError
```
[the remark this answers is not in the document]
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: Tis not my change , there is a handler set up for it in view selected survey which redirects to a default view of the messages. I guess that KeyError may be too broad to catch here and  is deliberatly left to perculate .

File: response_operations_ui/controllers/message_controllers.py [hunk 5]
```diff
@@ -40,11 +40,27 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
     logger.info("Retrieving count of threads",
                 survey_id=survey_id, conversation_tab=conversation_tab, business_id=business_id)
 
-    params = _get_secure_message_threads_params(survey_id, business_id, conversation_tab)
+    response = _get_conversation_counts(business_id, conversation_tab, survey_id, all_conversation_types=False)
+    try:
+        response.raise_for_status()
+    except HTTPError:
+        logger.exception("Thread count failed")
+        raise ApiError(response)
```
[the remark this answers is not in the document]
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: I think your right, in that  it should be part of a wider conversation of how do we deal with errors.  But in this specific case without a count we cant paginate and so cant give a meaningful display.  So I think its right to error as the user will be totally confused if not on the first page . In the case where there is only one page , which may be the case when getting all the counts then maybe you have a case, but even then I think its is adding complexity for an error case under a too precise a subset of conditions.

File: response_operations_ui/controllers/message_controllers.py [hunk 6]
```diff
@@ -40,11 +40,27 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
...
+    try:
+        return response.json()['total']
+    except KeyError:
+        logger.exception("Response was successful but didn't contain a 'total' key")
+        raise NoMessagesError
 
-    response = requests.get(url, headers={'Authorization': _get_jwt()}, params=params)
+
+def get_all_conversation_type_counts(survey_id, conversation_tab, business_id):
+    """gets the count for teh current tab and the count for the  conversations in all 4 tabs"""
```
[the remark this answers is not in the document]
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: Thanks

File: response_operations_ui/controllers/message_controllers.py [hunk 7]
```diff
@@ -53,19 +69,37 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
...
     logger.info("Count successful")
+
     try:
-        return response.json()['total']
+        totals = response.json()['totals']
+
+        # Secure Message uses different identifiers to the tab names used in the ui, this translates the names
+        totals['initial'] = totals.pop('new_respondent_conversations')
+        totals['my messages'] = totals.pop('my_conversations')
+
```
[the remark this answers is not in the document]
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: Dont follow ?

File: response_operations_ui/controllers/message_controllers.py [hunk 8]
```diff
@@ -53,19 +69,37 @@ def get_conversation_count(survey_id, conversation_tab, business_id):
...
 
     logger.info("Count successful")
+
     try:
-        return response.json()['total']
+        totals = response.json()['totals']
+
+        # Secure Message uses different identifiers to the tab names used in the ui, this translates the names
+        totals['initial'] = totals.pop('new_respondent_conversations')
+        totals['my messages'] = totals.pop('my_conversations')
```
[the remark this answers is not in the document]
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: It uses a space because during rendering it uses the tab title to select the value

File: response_operations_ui/controllers/message_controllers.py (hunk 1 above)
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: get leaves them in place with the incorrect name, pop replaces them . But there is a small  bug in that we will end up with initial and my messages being null in the case of getting a count for 1 tab. fiixing that.

File: response_operations_ui/controllers/message_controllers.py (hunk 2 above)
Comment (AndrewTorrance) [reply response_operations_ui/controllers/message_controllers.py]: After the change for the previous comment it can only be a missing totals key.

File: response_operations_ui/views/messages.py (hunk 3 above)
Comment (AndrewTorrance) [reply response_operations_ui/views/messages.py]: Good catch , I was passing the counts via session and this was initialising the dict. But I moved it off session and missed this . Thanks

File: response_operations_ui/controllers/message_controllers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (insacuri) [reply response_operations_ui/controllers/message_controllers.py]: ah fair enough
